File: examples/langchain_python/complex_agent.py
"""Complex LangChain agent with multiple tools and advanced configuration."""

from langchain.agents import initialize_agent, AgentType
from langchain.tools import tool
from langchain_openai import ChatOpenAI
from langchain.memory import ConversationBufferMemory


@tool
def web_search(query: str) -> str:
    """Search the web for information."""
    return f"Web search results for: {query}"


@tool
def send_email(recipient: str, subject: str, body: str) -> str:
    """Send an email to a recipient."""
    return f"Email sent to {recipient} with subject: {subject}"


@tool
def create_calendar_event(title: str, date: str, time: str) -> str:
    """Create a calendar event."""
    return f"Calendar event created: {title} on {date} at {time}"


@tool
def analyze_sentiment(text: str) -> str:
    """Analyze the sentiment of a given text."""
    return "Sentiment analysis: The text appears to be positive"


@tool
def summarize_text(text: str, max_words: int = 100) -> str:
    """Summarize a long text into a shorter version."""
    return f"Summary (max {max_words} words): Key points extracted"


@tool
def translate_text(text: str, target_language: str) -> str:
    """Translate text to a target language."""
    return f"Translation to {target_language}: [translated text]"


@tool
def extract_entities(text: str) -> str:
    """Extract named entities from text."""
    return "Entities found: Person: John Doe, Organization: Acme Corp"


# Initialize the language model with custom settings
llm = ChatOpenAI(model="gpt-4", temperature=0.3, max_tokens=2000)

# Initialize memory for conversation history
memory = ConversationBufferMemory(memory_key="chat_history", return_messages=True)

# Create the tools list
tools = [
    web_search,
    send_email,
    create_calendar_event,
    analyze_sentiment,
    summarize_text,
    translate_text,
    extract_entities,
]

# Initialize the complex agent
complex_agent = initialize_agent(
    tools=tools,
    llm=llm,
    agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION,
    agent_name="Executive Assistant Agent",
    verbose=True,
    memory=memory,
    handle_parsing_errors=True,
    max_iterations=10,
)

# Example usage
if __name__ == "__main__":
    response = complex_agent.run(
        "Search for information about AI trends, summarize the findings, and schedule a meeting"
    )
    print(response)

File: examples/langchain_typescript/complex_agent.ts
/**
 * Complex LangChain TypeScript agent with multiple tools
 */

import { AgentExecutor } from "langchain/agents";
import { ChatOpenAI } from "@langchain/openai";
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import { BufferMemory } from "langchain/memory";

// Define comprehensive tool set
const webSearch = tool({
  name: "web_search",
  description: "Search the web for information",
  schema: z.object({
    query: z.string().describe("Search query"),
  }),
  func: async ({ query }) => {
    return `Web search results for: ${query}`;
  },
});

const sendEmail = tool({
  name: "send_email",
  description: "Send an email to a recipient",
  schema: z.object({
    recipient: z.string().describe("Email recipient"),
    subject: z.string().describe("Email subject"),
    body: z.string().describe("Email body"),
  }),
  func: async ({ recipient, subject, body }) => {
    return `Email sent to ${recipient} with subject: ${subject}`;
  },
});

const createCalendarEvent = tool({
  name: "create_calendar_event",
  description: "Create a calendar event",
  schema: z.object({
    title: z.string().describe("Event title"),
    date: z.string().describe("Event date"),
    time: z.string().describe("Event time"),
  }),
  func: async ({ title, date, time }) => {
    return `Calendar event created: ${title} on ${date} at ${time}`;
  },
});

const analyzeSentiment = tool({
  name: "analyze_sentiment",
  description: "Analyze the sentiment of a given text",
  schema: z.object({
    text: z.string().describe("Text to analyze"),
  }),
  func: async ({ text }) => {
    return "Sentiment analysis: The text appears to be positive";
  },
});

const summarizeText = tool({
  name: "summarize_text",
  description: "Summarize a long text into a shorter version",
  schema: z.object({
    text: z.string().describe("Text to summarize"),
    maxWords: z.number().optional().describe("Maximum words in summary"),
  }),
  func: async ({ text, maxWords = 100 }) => {
    return `Summary (max ${maxWords} words): Key points extracted`;
  },
});

const translateText = tool({
  name: "translate_text",
  description: "Translate text to a target language",
  schema: z.object({
    text: z.string().describe("Text to translate"),
    targetLanguage: z.string().describe("Target language"),
  }),
  func: async ({ text, targetLanguage }) => {
    return `Translation to ${targetLanguage}: [translated text]`;
  },
});

const extractEntities = tool({
  name: "extract_entities",
  description: "Extract named entities from text",
  schema: z.object({
    text: z.string().describe("Text to extract entities from"),
  }),
  func: async ({ text }) => {
    return "Entities found: Person: John Doe, Organization: Acme Corp";
  },
});

// Initialize the language model with custom settings
const llm = new ChatOpenAI({
  modelName: "gpt-4",
  temperature: 0.3,
  maxTokens: 2000,
});

// Initialize memory
const memory = new BufferMemory({
  memoryKey: "chat_history",
  returnMessages: true,
});

// Create the tools array
const tools = [
  webSearch,
  sendEmail,
  createCalendarEvent,
  analyzeSentiment,
  summarizeText,
  translateText,
  extractEntities,
];

// Create the complex agent executor
const complexAgent = new AgentExecutor({
  tools: tools,
  agent: llm,
  verbose: true,
  memory: memory,
  maxIterations: 15,
});

// Example usage
async function main() {
  const result = await complexAgent.invoke({
    input: "Search for information about AI trends, summarize the findings, and schedule a meeting"
  });
  console.log(result);
}

if (require.main === module) {
  main();
}

export { complexAgent };

